### PR TITLE
Rename examples for ease of browsing

### PR DESCRIPTION
**Pull request by tsauvajon**
## Which problem is this PR solving?

Examples are hard to browse for newcomers, or when you want to see how you can use a Patron component specifically.
When someone wants to use gRPC with Patron, for example, they should be able to simply go to `examples` and find the `grpc` folder. Same thing for Kafka, SNS...

It also makes the interactions between service easier to understand :
- `third` consumes messages produced by `second`, and produce messages consumed by `fourth`
- `amqp` consumes messages produced by `kafka`, and produce messages consumed by `sns`

This helps with #84

## Short description of the changes

Renamed examples, new names are the main Patron component they use:
- `first` -> `http`
- `second` -> `kafka`
- `third` -> `amqp`
- `fourth` -> `sns`
- `fifth` -> `sqs`
- `sixth` -> `grpc`
- `seventh` -> `timing`, this could be named something different.

### CODE REVIEW

**Comment by mantzas** [comment]
@tsauvajon 
The idea behind the names is that we generate a chain of services: first->second->...->seventh.
So in order to run them you start from the last and run it and then the one before that and eventually all services are started.
The next step is to send a request to first and see this propagate to all the others...
We have also per service multiple things in there: example first is JSON HTTP and Protobuf HTTP client....
So naming them as propose is also not very clear...
Any idea how to handle this?

**Comment by tsauvajon** [comment]
@mantzas thanks for the review

> @tsauvajon
> The idea behind the names is that we generate a chain of services: first->second->...->seventh.
> So in order to run them you start from the last and run it and then the one before that and eventually all services are started.

I understand your point of view.

It was not obvious for me at first that these examples were runnable and that you had to run them in the correct order - much less that it had to be in reverse order (from `7th` to `1st`). Besides, I don't expect most people to actually run the examples - most would probably just want to have a look at how to use a specific component with Patron. And in any case, if I wanted to run the examples, then I would dig in and read the README, that clearly indicates all the steps on how to run the examples, instead of trying to figure out the relationships and dependencies by myself.

My point is that I would make browsing the examples as simple as possible, making the folders self-documented - not even having to read the README when looking for, say, a Kafka example. For users that want to dig in and see the examples running, then relying on the README for that is perfectly acceptable.

Note that I'm not saying that people should jump in and start reading code before reading the README, but they do! In my case, I think that code examples speaks better than documentation - at least it's easier for me to understand a framework that way.

----

> We have also per service multiple things in there: example first is JSON HTTP and Protobuf HTTP client....
> So naming them as propose is also not very clear...
> Any idea how to handle this?

For services with multiple things, I would simply name them with the Patron component they are using. If they contain more than one component, either:
- use the most significant component used for the example name
- use multiple component names (e.g. snssqs for SNS + SQS examples)
- make 1 example per component (they can be self-standing examples), on top of the existing ones.

For `first`, there's no explicit JSON serialisation/deserialisation nor import. I would say that the point of the example isn't to showcase `protobuf` encoding either. We could make mini examples showing JSON/protobuf encoding. And then either keep this example as the all-emcompassing `http` example, or make mini examples for just an HTTP server and another for just the HTTP client.

File: examples/README.md [hunk 1]
```diff
@@ -40,8 +40,8 @@ The processing will be kicked of by sending a request to the HTTP component. The
 
 - receives a raw http request
 - returns the 7th minute unix interval for the current server time
-- first makes a request to seventh
-- seventh responds with the timing information
+- http makes a request to the timing service
```
Comment (drakos74) [inline examples/README.md]: the test here is for the server cache. Maybe better to try use this term to connect conceptually with the related feature. Timing service is just a naive implementation of the server side cache

**Review by drakos74 — COMMENTED** [review]
nice idea! 
One small concern on the 'timing service'

File: examples/README.md (hunk 1 above)
Comment (tsauvajon) [reply examples/README.md]: `timing` is a very bad name indeed. Would you suggest `cache`, `server-cache`?

**Comment by mantzas** [comment]
> @mantzas thanks for the review
> 
> > @tsauvajon
> > The idea behind the names is that we generate a chain of services: first->second->...->seventh.
> > So in order to run them you start from the last and run it and then the one before that and eventually all services are started.
> 
> I understand your point of view.
> 
> It was not obvious for me at first that these examples were runnable and that you had to run them in the correct order - much less that it had to be in reverse order (from `7th` to `1st`). Besides, I don't expect most people to actually run the examples - most would probably just want to have a look at how to use a specific component with Patron. And in any case, if I wanted to run the examples, then I would dig in and read the README, that clearly indicates all the steps on how to run the examples, instead of trying to figure out the relationships and dependencies by myself.
> 
> My point is that I would make browsing the examples as simple as possible, making the folders self-documented - not even having to read the README when looking for, say, a Kafka example. For users that want to dig in and see the examples running, then relying on the README for that is perfectly acceptable.
> 
> Note that I'm not saying that people should jump in and start reading code before reading the README, but they do! In my case, I think that code examples speaks better than documentation - at least it's easier for me to understand a framework that way.
> 
> > We have also per service multiple things in there: example first is JSON HTTP and Protobuf HTTP client....
> > So naming them as propose is also not very clear...
> > Any idea how to handle this?
> 
> For services with multiple things, I would simply name them with the Patron component they are using. If they contain more than one component, either:
> 
>     * use the most significant component used for the example name
> 
>     * use multiple component names (e.g. snssqs for SNS + SQS examples)
> 
>     * make 1 example per component (they can be self-standing examples), on top of the existing ones.
> 
> 
> For `first`, there's no explicit JSON serialisation/deserialisation nor import. I would say that the point of the example isn't to showcase `protobuf` encoding either. We could make mini examples showing JSON/protobuf encoding. And then either keep this example as the all-emcompassing `http` example, or make mini examples for just an HTTP server and another for just the HTTP client.

then let's name them based on the "default" component that they use....

**Comment by tpaschalis** [comment]
Just to clear a small misunderstanding, once you've `docker-compose up -d` you can `go run` the examples in any order, try for yourself! You just have to send a request to the 'first' service as an entrypoint to get them working.

Their enumeration helps provide a mental model of how the requests flow, as you can see in the following diagram from Jaeger.

I don't have a strong opinion on how they should be named (eg. if `second-kafka` looks good or not), but in any case, I personally believe that if we remove this enumeration, we could write a sentence or two to explain how they piece together.

image

**Review by hugorut — DISMISSED** [review]
I'm in favour of this approach. I too found the examples folder incredibly confusing with the naming. We already have notes about the enumeration and how to run them in sequence in the examples `readme` so I think calling the examples numeric names is overkill. Providing the naming as @tsauvajon has done allows developers to jump into components they are interested in, straight from github if need be. I rarely run examples like these on my local machine, instead choosing to read to understand the code.

File: examples/README.md (hunk 1 above)
Comment (mantzas) [reply examples/README.md]: `http-cache` maybe?

Comment (tsauvajon) [reply examples/README.md]: `http-cache` makes it super obvious, I like it

**Review by mantzas — DISMISSED** [review]

**Review by tpaschalis — DISMISSED** [review]

**Comment by mantzas** [comment]
@tsauvajon hey, somebody beat you to the punch... there are some conflicts...

File: examples/README.md [hunk 2]
```diff
@@ -70,13 +70,13 @@ When the services started with Docker Compose are ready, you will need to start
...
-go run examples/fourth/main.go
-go run examples/fifth/main.go
-go run examples/sixth/main.go examples/sixth/greeter.pb.go
+go run examples/http-cache/main.go
+go run examples/http/main.go
+go run examples/kafka/main.go
+go run examples/amqp/main.go
+go run examples/sns/main.go
+go run examples/sqs/main.go
+go run examples/grpc/main.go examples/grpc/greeter.pb.go
```
Comment (tpaschalis) [inline examples/README.md:79]: (Sorry in advance for the nitpicky comment)

There's actually no `examples/grpc/greeter.pb.go` path. There is a `./component/grpc/greeter/greeter.pb.go`, but I'm not sure if it's required to run the examples.

**Review by mantzas — APPROVED** [review]

File: examples/README.md (hunk 2 above)
Comment (tpaschalis) [reply examples/README.md:79]: > I felt like it wasn't something to do as part of this PR and intentionally left it out.

Okay, I see your point. I'll try to run the examples and see if we need another issue.

**Review by tpaschalis — APPROVED** [review]

**Comment by tsauvajon** [comment]
@mantzas @tpaschalis what's the next step before we merge this PR?

File: examples/README.md (hunk 2 above)
Comment (tsauvajon) [reply examples/README.md:79]: I also had issues with the SNS and SQS examples (it wouldn't create the queues) and had a conflict with two examples trying to run on the same port.

**Comment by tpaschalis** [comment]
> @mantzas @tpaschalis what's the next step before we merge this PR?

Nothing's left to be done, I'm merging it as we speak! Thanks for the contribution, I think it will make navigating the examples easier.

File: examples/README.md (hunk 2 above)
Comment (tpaschalis) [reply examples/README.md:79]: Thanks for the feedback! The SNS/SQS examples should be okay now at `master (fixed by #282); and the ports conflict is being tracked by #277. The examples should be a-okay after the latter is resolved!

**Comment by mantzas** [comment]
@tsauvajon @tpaschalis @drakos74 The names are unfortunately wrong e.g. AMQP example is actually Kafka. This will be fixed with the documentation overhaul...

**Pull request merged**